### PR TITLE
NMS-13990: DeviceConfigBackup Rest API for downloading config data

### PR DESCRIPTION
**Pull request by synqotik**
Rest API for downloading the actual device config backup configuration data.

Pass comma-separated `id` query parameters with the `device_config` id.

- `GET /rest/device-config/download?id=1` downloads a single file, with encoding as per the file (for now, `text/plain`)
- `GET /rest/device-config/download?id=1,2,3` downloads a single `.tar.gz` file with the configuration files for ids 1, 2 and 3 inside.

Incorporated tar/gzip helpers in `DeviceConfigUtil` class.

Jira: https://issues.opennms.org/browse/NMS-13990

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/13990

### CODE REVIEW

File: features/device-config/persistence/api/src/main/java/org/opennms/features/deviceconfig/persistence/api/DeviceConfig.java [hunk 1]
```diff
@@ -105,9 +105,7 @@ public void setId(Long id) {
         this.id = id;
     }
 
-    public OnmsIpInterface getIpInterface() {
-        return ipInterface;
-    }
+    public OnmsIpInterface getIpInterface() { return ipInterface; }
```
Comment (cgorantla) [inline features/device-config/persistence/api/src/main/java/org/opennms/features/deviceconfig/persistence/api/DeviceConfig.java:108]: no need to touch this file

File: features/device-config/rest/pom.xml [hunk 2]
```diff
@@ -86,16 +80,26 @@
...
+            <artifactId>hamcrest</artifactId>
+            <version>2.2</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.opennms.features.device-config</groupId>
+            <artifactId>org.opennms.features.device-config.api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
         <dependency>
```
Comment (cgorantla) [inline features/device-config/rest/pom.xml:94]: test dependencies should be grouped together below main dependencies

File: features/device-config/rest/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <blueprint
-  xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
```
Comment (cgorantla) [inline features/device-config/rest/src/main/resources/OSGI-INF/blueprint/blueprint.xml:2]: no need to touch this file

**Review by cgorantla — COMMENTED** [review]
Looks good to me codewise, minor comments.
Also please resolve merge conflicts, will validate this once locally.

File: features/device-config/api/pom.xml [hunk 4]
```diff
@@ -29,4 +29,30 @@
...
     </build>
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-compress</artifactId>
+            <version>1.21</version>
```
[the remark this answers is not in the document]
Comment (synqotik) [reply features/device-config/api/pom.xml]: Fixed

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigUtil.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+        return result;
+    }
+
+    public static byte[] tarMultipleFiles(Map<String, byte[]> fileNameToDataMap) throws IOException {
+        var byteOutputStream = new ByteArrayOutputStream();
+
+        try (var tarOut = new TarArchiveOutputStream(byteOutputStream)) {
+            for (Map.Entry<String, byte[]> entry : fileNameToDataMap.entrySet()) {
+                if (!Strings.isNullOrEmpty(entry.getKey()) && ArrayUtils.isNotEmpty(entry.getValue())) {
```
[the remark this answers is not in the document]
Comment (synqotik) [reply features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigUtil.java]: Fixed to just a null check

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigUtil.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        try (var gzipOut = new GzipCompressorOutputStream(byteOutputStream)) {
+            gzipOut.write(tarBytes);
+        }
+
+        return byteOutputStream.toByteArray();
+    }
+
+    public static Map<String, byte[]> unTarGzipMultipleFiles(byte[] tarGzipData) throws IOException {
+        if (ArrayUtils.isEmpty(tarGzipData)) {
+            return null;
```
[the remark this answers is not in the document]
Comment (synqotik) [reply features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigUtil.java]: Fix to just a null check, returns empty HashMap

File: features/device-config/persistence/api/src/main/java/org/opennms/features/deviceconfig/persistence/api/DeviceConfig.java (hunk 1 above)
Comment (synqotik) [reply features/device-config/persistence/api/src/main/java/org/opennms/features/deviceconfig/persistence/api/DeviceConfig.java]: Reverted - I think IntelliJ did this ;)

File: features/device-config/rest/pom.xml (hunk 2 above)
Comment (synqotik) [reply features/device-config/rest/pom.xml:68]: Fixed

File: features/device-config/rest/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 3 above)
Comment (synqotik) [reply features/device-config/rest/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Reverted

**Review by cgorantla — APPROVED** [review]

**Pull request merged**